src/container/srv_target.c

```diff
@@ -2104,3 +2104,230 @@ ds_cont_oid_alloc_handler(crt_rpc_t *rpc)
 	out->co_map_version = 0;
 	crt_reply_send(rpc);
 }
+
+struct cont_ec_eph {
+	uuid_t		cont_uuid;
+	daos_epoch_t	eph;
+	uint32_t	updated:1;
+};
+
+struct cont_ec_xs_query_arg {
+	uuid_t			pool_uuid;
+	int			tgt_id;
+	int			ephs_cnt;
+	int			ephs_alloc_cnt;
+	struct cont_ec_eph	*ephs;
+};
+
+static int
+cont_ec_xs_reduce_alloc(struct dss_stream_arg_type *xs, void *agg_arg)
+{
+	struct cont_ec_xs_query_arg *xs_arg;
+	struct cont_ec_xs_query_arg *a_arg = agg_arg;
+
+	D_ALLOC_PTR(xs_arg);
+	if (xs_arg == NULL)
+		return -DER_NOMEM;
+
+	uuid_copy(xs_arg->pool_uuid, a_arg->pool_uuid);
+	xs->st_arg = xs_arg;
+	return 0;
+}
+
+static void
+cont_ec_xs_reduce_free(struct dss_stream_arg_type *xs)
+{
+	struct cont_ec_xs_query_arg *xs_arg = xs->st_arg;
+
+	D_FREE_PTR(xs_arg->ephs);
+	D_FREE_PTR(xs_arg);
+}
+
+struct cont_ec_eph *
+lookup_insert_cont_ec_eph(struct cont_ec_xs_query_arg *arg, uuid_t cont_uuid)
+{
+	struct cont_ec_eph *found = NULL;
+	int i;
+
+	for (i = 0; i < arg->ephs_cnt; i++) {
+		if (uuid_compare(arg->ephs[i].cont_uuid, cont_uuid) == 0) {
+			found = &arg->ephs[i];
+			return found;
+		}
+	}
+
+	D_ASSERT(arg->ephs_cnt < arg->ephs_alloc_cnt);
+
+	uuid_copy(arg->ephs[arg->ephs_cnt].cont_uuid, cont_uuid);
+	found = &arg->ephs[arg->ephs_cnt];
+	arg->ephs_cnt++;
+	return found;
+}
+
+static void
+cont_ec_eph_reduce(void *agg_arg, void *xs_arg)
+{
+	struct cont_ec_xs_query_arg	*a_arg = agg_arg;
+	struct cont_ec_xs_query_arg	*x_arg = xs_arg;
+	int				i;
+
+	if (a_arg->ephs_alloc_cnt < x_arg->ephs_cnt) {
```

Comment:
How can this happen? Some VOS container is stopped on certain xstream? In that case, we need to make sure abort EC epoch updating for that container, right?

Reply:
well, on the 1st call, alloc_cnt is 0.

Reply:
Then it's better to change it from "if (a->cnt < x_cnt) realloc" to  "if (a->cnt == 0) alloc".